flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -0,0 +1,183 @@
+			double updateValue = 0;
+
+			for (double element : inMessages) {
+				if (getSuperstepNumber() == maxIteration) {
+					updateValue = element;
+					break;
+				}
+				updateValue += element;
+			}
+
+			// in the first iteration, no aggregation to call, init sum with value of vertex
+			double iterationValueSum = 1.0;
+
+			if (getSuperstepNumber() > 1) {
+				iterationValueSum = Math.sqrt(((DoubleValue) getPreviousIterationAggregate("sumVertexValue")).getValue());
+			}
+			if (getSuperstepNumber() != maxIteration) {
+				if (getSuperstepNumber() % 2 == 1) {
+					setNewVertexValue(new Tuple2<Double, Double>(vertex.getValue().f0 / iterationValueSum, updateValue));
+					doubleSumAggregator.aggregate(Math.pow(updateValue, 2));
+				} else {
+					setNewVertexValue(new Tuple2<Double, Double>(updateValue, vertex.getValue().f1 / iterationValueSum));
+					doubleSumAggregator.aggregate(Math.pow(updateValue, 2));
+				}
+			} else {
+
+				//final iteration to normalize hub score
+				setNewVertexValue(new Tuple2<Double, Double>(vertex.getValue().f0 / iterationValueSum, vertex.getValue().f1));
+			}
+		}
+	}
+
+	/**
+	 * Distributes the value of a vertex among all neighbor vertices and sum all the
+	 * value in every superstep.
+	 */
+	@SuppressWarnings("serial")
+	public static final class MessageUpdate<K, EV> extends MessagingFunction<K, Tuple2<Double, Double>, Double, EV> {
+		private int maxIteration;
+
+		public MessageUpdate(int maxIteration) {
+			this.maxIteration = maxIteration;
+		}
+
+		@Override
+		public void sendMessages(Vertex<K, Tuple2<Double, Double>> vertex) {
+
+			// in the first iteration, no aggregation to call, init sum with value of vertex
+			double iterationValueSum = 1.0;
+
+			if (getSuperstepNumber() > 1) {
+				iterationValueSum = Math.sqrt(((DoubleValue) getPreviousIterationAggregate("sumVertexValue")).getValue());
+			}
+			for (Edge<K, EV> edge : getEdges()) {
+				K messageSource = getSuperstepNumber() % 2 == 1 ? edge.getSource() : edge.getTarget();
+				K messageTarget = getSuperstepNumber() % 2 == 1 ? edge.getTarget() : edge.getSource();
+				Double messageValue = getSuperstepNumber() % 2 == 1 ? vertex.getValue().f0 : vertex.getValue().f1;
+
+				if (!messageTarget.equals(vertex.getId())) {
+					if (getSuperstepNumber() != maxIteration) {
+						sendMessageTo(messageTarget, messageValue / iterationValueSum);
+
+						// in order to make every vertex updated
+						sendMessageTo(messageSource, 0.0);
+					} else {
+						sendMessageTo(messageSource, iterationValueSum);
+					}
+				}
+			}
+		}
+	}
+
+	public static class VertexInitMapper<K, VV> implements MapFunction<Vertex<K, VV>, Tuple2<Double, Double>> {
+
+		private static final long serialVersionUID = 1L;
+
+		public Tuple2<Double, Double> map(Vertex<K, VV> value) {
+			
+			//init hub and authority value of each vertex
+			return new Tuple2<Double, Double>(1.0, 1.0);
```

Comment:
This object can be created as a member variable so that only a single object is instantiated.